sdk/src/main/java/com/kin/ecosystem/recovery/BackupAndRestoreImpl.java

```diff
@@ -0,0 +1,127 @@
+import com.kin.ecosystem.core.bi.EventLogger;
+import com.kin.ecosystem.core.bi.events.BackupWalletCompleted;
+import com.kin.ecosystem.core.bi.events.GeneralEcosystemSdkError;
+import com.kin.ecosystem.core.bi.events.RestoreWalletCompleted;
+import com.kin.ecosystem.core.data.blockchain.BlockchainSource;
+import com.kin.ecosystem.core.data.settings.SettingsDataSource;
+import com.kin.ecosystem.core.util.Validator;
+
+public class BackupAndRestoreImpl implements BackupAndRestore {
+
+	protected final BackupManager backupManager;
+	private final AccountManager accountManager;
+	private final EventLogger eventLogger;
+	private final SettingsDataSource settingsDataSource;
+	private final BlockchainSource blockchainSource;
+
+	public BackupAndRestoreImpl(@NonNull final Activity activity, @NonNull AccountManager accountManager,
+		@NonNull EventLogger eventLogger, @NonNull BlockchainSource blockchainSource, SettingsDataSource settingsDataSource) {
+		this.blockchainSource = blockchainSource;
+		this.backupManager = new BackupManager(activity, blockchainSource.getKeyStoreProvider());
+		this.accountManager = accountManager;
+		this.eventLogger = eventLogger;
+		this.settingsDataSource = settingsDataSource;
+	}
+
+	@Override
+	public void backupFlow() throws ClientException {
+		if (blockchainSource.getKinAccount() != null) {
+			backupManager.backupFlow();
+		} else {
+			throw new ClientException(ClientException.ACCOUNT_NOT_LOGGED_IN, "Account should be logged in before backup.", null);
+		}
+	}
+
+	@Override
+	public void restoreFlow() {
+		backupManager.restoreFlow();
+	}
+
+	@Override
+	public void registerBackupCallback(@NonNull final BackupAndRestoreCallback backupCallback) {
+		Validator.checkNotNull(backupCallback, "backupCallback");
+		backupManager.registerBackupCallback(new BackupCallback() {
+
+			@Override
+			public void onSuccess() {
+				try {
+					final String publicAddress = blockchainSource.getPublicAddress();
+					settingsDataSource.setIsBackedUp(publicAddress, true);
+				} catch (ClientException | BlockchainException e) {
+					eventLogger.send(GeneralEcosystemSdkError
+						.create("BackupAndRestoreImpl onSuccess blockchainSource.getPublicAddress() thrown an exception"));
+				}
+				eventLogger.send(BackupWalletCompleted.create());
+				backupCallback.onSuccess();
+			}
+
+			@Override
+			public void onCancel() {
+				backupCallback.onCancel();
+			}
+
+			@Override
+			public void onFailure(Throwable throwable) {
+				//TODO create our exceptions
+				backupCallback.onFailure(throwable);
+			}
+		});
+	}
+
+	private void switchAccount(int accountIndex, @NonNull final BackupAndRestoreCallback backupCallback) {
+		accountManager.switchAccount(accountIndex, new KinCallback<Boolean>() {
+			@Override
+			public void onResponse(Boolean response) {
+				eventLogger.send(RestoreWalletCompleted.create());
+				backupCallback.onSuccess();
+			}
+
+			@Override
+			public void onFailure(KinEcosystemException exception) {
```

Comment:
does switch account know to rollback in case of error?

Reply:
we are updating the current account only if switch account with our servers response success.

Reply:
👍